### ISSUE TITLE
Single vaccine

### ISSUE THREAD

**Issue by giovannic** [mrc-ide/hypatia#21]
Todo:

- [ ] Parameterise a vaccine
- [ ] Parameterise a vaccine strategy
- [ ] Create variable to track time of last vaccination(s)
- [ ] Create vaccine distribution process
- [ ] Modify infection process xply prob_infection and prob_severe by vaccine efficacy for each individual

**Comment by pwinskill** [comment]
Nimue currently has vaccinated individuals going through 2 compartments before losing vaccine protection (if duration of vaccine derived immunity is <Inf). We assume immunity last for the longest period determined by natural expsoure (if it has occured) or vaccination.

I'm wondering if we want to model each vaccine with a very flexible set of (age-dependent) characteristics that could include any combination of the following:

- Infection blocking
- Reduction in hospitalisation rate
- Reduction in onwards transmission (independent to reductions in durations/infection etc?).
- Reduction in duration of severe illness
- Reduction in duration of mild illness
- Increase in the proportion of mild cases that are symptomatic

**Comment by pwinskill** [comment]
Some of the above, may be version 3 considerations, but it might be easier to code the at the same time with null defaults

**Comment by pwinskill** [comment]
After futher discussion with OJ, proposed intial appraoch would be to include a flexible vaccine parameterisation, first to capture reductions in the probability of infection and/or the probability of severe case.

This would be implemented by allowing the user to specify an efficacy curve (for each mode of action) as a look up table. When implementing, for vaccinated individuals we will estimate the time since fully vaccinated (`vx_t`) and use that value to look up the vaccine-impact from the appropriate table that will then be used to modify `prob_infection` or `prob_severe`

**Comment by giovannic** [comment]
Sweet (I've put these tasks in the header)

I still have some questions around distribution strategy. Are we gonna do like nimue?

*Parameterise vaccine*

We need to take:

 * efficacy_curve, a vector of efficacy values for an arbitrary number of timesteps after vaccination 

*Parameterise vaccine strategy* (I'm guessing here)

We need to take:

 * prob_vaccination, the probability of each age group getting the vaccination
 * n_vaccine, the number of available vaccines available at `tt_vaccine`
 * tt_vaccine, the timesteps when `n_vaccine` become available

**Comment by pwinskill** [comment]
Yes, I think the distribution part may be the trickiest bit. I think some aspects we should retain in the nimue style, namely, for each vaccine implemented:
1. Allow the user to define a population level, time varying vaccination rate: doses per unit time
2. Define a prioritisation matrix - I think the nimue approach of a age X prioritisation step coverage matrix works well
3. Translate 1 and 2 to vaccination of individuals in appropriate age groups at appropriate times in the IBM

**Comment by giovannic** [comment]
Vaccine allocation:

 * variable for prioritisation block (how do we assign this?)
  Perhaps with vectors for:
    * age ranges
    * occupation
    * time period
  (tbc by @pwinskill )

 * Exclude people who are in hospital or sick

Subsequent doses:
 * Must count towards total available doses
 * Deterministic delay time for booster doses
 * Strategy allows you to decide if subsequent doses are affected by sickness
 * There is a dropout based on bernoulli draws

**Comment by giovannic** [comment]
Ok so. For allocation we should have:

 * We should be able to parameterise eligibility over time. Eligibility can be based on:
   * age range
   * hospital worker (yes/no) - these variables have to be sampled from data that will soon arrive in nimue
   * has comorbidites (yes/no) - these variables have to be sampled from data that will soon arrive in nimue

There will be an age based coverage/hesitancy which shouldn't be exceeded.

Each timestep, eligible individuals will be uniformly allocated doses (maintaining coverage and total availablity)

**Comment by pwinskill** [comment]
We might want to include:

- Vaccine eligable and/or vaccine hesitant

as well. This would allow us to capture those who cannot or will not have the vaccine
